src/file_system.rs

```diff
@@ -233,31 +190,96 @@ fn metadata() {
     let _ = meta;
 }
 
-fn node_compatible_raw_canonicalize<P: AsRef<Path>>(path: P) -> io::Result<PathBuf> {
-    cfg_if! {
-        if #[cfg(windows)] {
-            use std::path::{Component, Prefix};
-            // same logic with https://github.com/libuv/libuv/blob/d4ab6fbba4669935a6bc23645372dfe4ac29ab39/src/win/fs.c#L2774-L2784
-            let canonicalized = fs::canonicalize(path)?;
-            let first_component = canonicalized.components().next();
-            match first_component {
-                Some(Component::Prefix(prefix)) => {
-                    match prefix.kind() {
-                        Prefix::VerbatimUNC(_, _) => {
-                            Ok(canonicalized.to_str().and_then(|s| s.get(UNC_PATH_PREFIX.len()..)).map(PathBuf::from).unwrap_or(canonicalized))
-                        }
-                        Prefix::VerbatimDisk(_) => {
-                            Ok(canonicalized.to_str().and_then(|s| s.get(LONG_PATH_PREFIX.len()..)).map(PathBuf::from).unwrap_or(canonicalized))
-                        }
-                        _ => {
-                            Ok(canonicalized)
-                        }
+#[inline]
+// This is A faster fs::canonicalize implementation by reducing the number of syscalls
+fn fast_canonicalize(path: &Path) -> io::Result<PathBuf> {
+    use std::path::Component;
+    let mut path_buf = path.to_path_buf();
+
+    loop {
+        let link = {
+            #[cfg(windows)]
+            {
+                node_compatible_raw_canonicalize(fs::read_link(&path_buf)?)
+            }
+            #[cfg(not(windows))]
+            {
+                fs::read_link(&path_buf)?
+            }
+        };
+        path_buf.pop();
+        if fs::symlink_metadata(&path_buf)?.is_symlink() {
```

Comment:
It's not needed for this PR, but writing down an idea. I guess we can reduce syscalls further more by caching `fs::symlink_metadata`.

Reply:
Let's do it later 😁